File: Python/Leetcode_Python/BinarySearch/SearchRotatedArray.py
# Time Complexity: O(log n) || Space Complexity: O(1)
from typing import List


def search(nums: List[int], target: int) -> int:
    if not nums:
        return -1
    if len(nums) == 1 and target != nums[0]:
        return -1
    if len(nums) == 1 and target == nums[0]:
        return 0

    start = 0
    end = len(nums) - 1
    while start < end:
        mid = (start + end) // 2
        if nums[start] == target:
            return start
        if nums[end] == target:
            return end
        if nums[mid] == target:
            return mid

        if (
            (target < nums[mid] and nums[mid] < nums[end])
            or (target > nums[mid] and target > nums[end] and nums[mid] < nums[end])
            or (target < nums[mid] and target > nums[end])
        ):
            end = mid - 1
        else:
            start = mid + 1

    return -1

File: Java/Leetcode/BinarySearch/SearchRotatedArray.java
package com.brianprojects.Leetcode.BinarySearch;
public class SearchRotatedArray {
    public int search(int[] nums, int target) {
        int end = nums.length-1;
        int start = 0;

        while(start < end){
            int mid = (start+end)/2;
            if(target == nums[mid]){
                return mid;
            }

            if(nums[start] <= nums[mid]){
                if(target >= nums[start] && target < nums[mid]){
                    end = mid-1;
                }
                else{
                    start = mid+1;
                }
            }
            else{
                if(target > nums[mid] && target <=nums[end]){
                    start = mid+1;
                }
                else{
                    end = mid-1;
                }
            }
        }
        return nums[start] == target? start: -1;
    }


}
